src/Workspaces/Core/Portable/Storage/CloudCache/CloudCachePersistentStorage.cs

```diff
@@ -0,0 +1,194 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.IO;
+using System.IO.Pipelines;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.PersistentStorage;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.Storage
+{
+    /// <summary>
+    /// Implementation of Roslyn's <see cref="IPersistentStorage"/> sitting on top of the platform's cloud storage
+    /// system.
+    /// </summary>
+    internal class CloudCachePersistentStorage : AbstractPersistentStorage
+    {
+        private static readonly ObjectPool<byte[]> s_byteArrayPool = new(() => new byte[Checksum.HashSize]);
+
+        /// <remarks>
+        /// We do not need to store anything specific about the solution in this key as the platform cloud cache is
+        /// already keyed to the current solution.  So this just allows us to store values considering that as the root.
+        /// </remarks>
+        private static readonly RoslynCloudCacheContainerKey s_solutionKey = new("Roslyn.Solution");
+
+        /// <summary>
+        /// Cache from project green nodes to the container keys we've computed for it (and the documents inside of it).
+        /// We can avoid computing these container keys when called repeatedly for the same projects/documents.
+        /// </summary>
+        private static readonly ConditionalWeakTable<ProjectState, ProjectContainerKeyCache> s_projectToContainerKey = new();
+        private readonly ConditionalWeakTable<ProjectState, ProjectContainerKeyCache>.CreateValueCallback _projectToContainerKeyCallback;
+
+        /// <summary>
+        /// Underlying cache service (owned by platform team) responsible for actual storage and retrieval of data.
+        /// </summary>
+        private readonly IRoslynCloudCacheService _cacheService;
+
+        public CloudCachePersistentStorage(
+            SolutionKey solutionKey,
+            IRoslynCloudCacheService cacheService,
+            string workingFolderPath,
+            string relativePathBase,
+            string databaseFilePath)
+            : base(workingFolderPath, relativePathBase, databaseFilePath)
+        {
+            _cacheService = cacheService;
+            _projectToContainerKeyCallback = ps => new(relativePathBase, ProjectKey.ToProjectKey(solutionKey, ps));
```

Comment:
Since you asked: this was a good example where the target-typed new was icky to review, the only way to know what this is creating is to go back and look at the field and infer it yourself; the naming confusion (which you since fixed) meant it was really confusing if this was creating a container key or something else.

Reply:
will unicky this.